### PR TITLE
Add JSON.sh

### PR DESCRIPTION
**Pull request by karlding**
This fixes all the ShellCheck errors as per #48

### CODE REVIEW

File: lib/JSON.sh [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+  esac
+}
+
+if ([ "$0" = "${BASH_SOURCE[0]}" ] || ! [ -n "${BASH_SOURCE[0]}" ]);
+then
+  parse_options "$@"
+  tokenize | parse
+fi
+
+# vi: expandtab sw=2 ts=2
```
Comment (Hazematman) [inline lib/JSON.sh:215]: Comment not relevant to code?

**Review by Hazematman — COMMENTED** [review]
check comments

File: lib/JSON.sh (hunk 1 above)
Comment (karlding) [reply lib/JSON.sh:215]: That's part of the original code. I'd rather not deviate too far from the original if it isn't necessary, in  case we need to integrate change from upstream.

With that being said, I can remove this if you really want to.

Comment (Hazematman) [reply lib/JSON.sh:215]: Nah keeping it like upstream is probably the best idea

**Review by Hazematman — APPROVED** [review]

**Comment by karlding** [comment]
@Caelink do you want to take a look before I merge?

File: lib/JSON.sh [hunk 2]
```diff
@@ -97,9 +97,13 @@ tokenize () {
 
   # Force zsh to expand $A into multiple words
   local is_wordsplit_disabled=$(unsetopt 2>/dev/null | grep -c '^shwordsplit$')
-  if [ $is_wordsplit_disabled != 0 ]; then setopt shwordsplit; fi
+  if [ "$is_wordsplit_disabled" != 0 ]; then
```
Comment (Caelink) [inline lib/JSON.sh]: double braces?

File: lib/JSON.sh [hunk 3]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
```
Comment (Caelink) [inline lib/JSON.sh]: At some point we should make this consistent in every file (some use !/usr/bin/env bash I believe?)

File: lib/JSON.sh [hunk 4]
```diff
@@ -203,7 +205,7 @@ parse () {
   esac
 }
 
-if ([ "$0" = "$BASH_SOURCE" ] || ! [ -n "$BASH_SOURCE" ]);
+if ([ "$0" = "${BASH_SOURCE[0]}" ] || ! [ -n "${BASH_SOURCE[0]}" ]);
```
Comment (Caelink) [inline lib/JSON.sh]: I think this should be another thing we change everything to be consistent on

File: lib/JSON.sh [hunk 5]
```diff
@@ -178,6 +178,7 @@ parse_value () {
     ''|[!0-9]) throw "EXPECTED value GOT ${token:-EOF}" ;;
     *) value=$token
        # if asked, replace solidus ("\/") in json strings with normalized value: "/"
+       # shellcheck disable=SC2001
```
Comment (Caelink) [inline lib/JSON.sh]: I think sed is fine so this seems good

**Review by Caelink — APPROVED** [review]
Minor comments about consistency and things that are not necessarily relevant here but are elsewhere in the codebase

File: lib/JSON.sh [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+  if echo "test string" | grep -Eao --color=never "test" >/dev/null 2>&1
+  then
+    GREP='egrep -ao --color=never'
+  else
+    GREP='egrep -ao'
+  fi
+
+  if echo "test string" | grep -Eo "test" >/dev/null 2>&1
+  then
```
Comment (Caelink) [inline lib/JSON.sh:84]: change if format file wide pls

**Pull request merged**